### DISCUSSION TITLE
Replace firefox's pdf reader with doqment

### DISCUSSION THREAD

**Discussion by KayJay7** [Floorp-Projects/Floorp#1214] [Feature Request]
The [doqment](https://github.com/shivaprsd/doqment) project offers a custom fork of mozilla's pdf.js used in firefox with many useful features, like changing the document colors.

They deploy it  as an add on, but on firefox (and chrome), the pdf viewers are treated differently from other pages and extension. Basically, extensions cannot affect the pdf reader, and when an extension provides a substitute reader (like in doqment's case), that reader doesn't have the same privileges as the native one. 

Some limitations are: it cannot open local files unless the user explicitly drag and drops them on the page, this includes tabs that where open in previous sessions, and the searchbar provided by the extension cannot be used and is replaced by firefox's own bar.
Overall, the limitations (especially the local files one) makes using the addon very frustrating, and one might rather use the less featured integrated viewer just for that.

If doqment was included as part of the browser, instead of an addon, in place of the original pdf.js, it wouldn't have the same limitations, and it would be a big improvement over the original viewer.

**Comment by shivaprsd** [comment]
Hello, author of *doqment* here. @KayJay7 I am glad to see your appreciation of *doqment* here. I understand your difficulties with the extension's limitations. I really appreciate your proposal to integrate it directly in Floorp. Let me provide some corrections/additional info for the Floorp devs if they are interested in this.

Firstly, *doqment* does not use a "custom fork" of PDF.js. It uses the [public build](https://github.com/mozilla/pdf.js/releases) of the generic viewer, with a [minuscule modification](https://github.com/shivaprsd/pdfjs-generic/compare/stable...main.diff). All the additional features are implemented on top by a set of add-on scripts. The main add-on module is [*doq*](https://github.com/shivaprsd/doq) (also written by me) which adds the ability to change the PDF colours.

This means that the base viewer that comes from FF source can be retained, and updated separately, from the add-on code. I wrote *doq* as a standalone library for any PDF.js deployment so that it is not tied to the extension. The other add-on script can be also be moved to a separate library repo if needed. I try to keep them all working with the latest updates of the PDF.js viewer.

I don't suggest "replacing" the built in viewer with *doqment* as the latter uses the generic build, while the former uses the Firefox-specific build. Some of the limitations (like search bar) is due to the build, not the extension. *doqment* also includes a lot of extension-specific code (for re-routing PDF traffic, setting options, etc) which are unwanted in an integrated viewer.

If the devs are interested in including the add-on modules only on the built-in viewer (and allow enabling them preferably behind an `about:config` flag), I would be happy to discuss and assist. I am also interested to see how my add-ons perform with other builds of PDF.js. 🙂
